src/Uno.UI/UI/Xaml/ResourceResolver.cs

```diff
@@ -0,0 +1,299 @@
+			{
+				return t;
+			}
+
+			return default(T);
+		}
+
+		/// <summary>
+		/// Push a new <see cref="XamlScope"/>, typically because a template is being materialized.
+		/// </summary>
+		/// <param name="scope"></param>
+		internal static void PushNewScope(XamlScope scope) => _scopeStack.Push(scope);
+		/// <summary>
+		/// Push a new Resources source to the current xaml scope.
+		/// </summary>
+		internal static void PushSourceToScope(DependencyObject source) => PushSourceToScope((source as IWeakReferenceProvider).WeakReference);
+		/// <summary>
+		/// Push a new Resources source to the current xaml scope.
+		/// </summary>
+		internal static void PushSourceToScope(ManagedWeakReference source)
+		{
+			var current = _scopeStack.Pop();
+			_scopeStack.Push(current.Push(source));
+		}
+		/// <summary>
+		/// Pop Resources source from current xaml scope.
+		/// </summary>
+		internal static void PopSourceFromScope()
+		{
+			var current = _scopeStack.Pop();
+			_scopeStack.Push(current.Pop());
+		}
+		/// <summary>
+		/// Pop current <see cref="XamlScope"/>, typically because template materialization is complete.
+		/// </summary>
+		internal static void PopScope()
+		{
+			_scopeStack.Pop();
+			if (_scopeStack.Count == 0)
+			{
+				throw new InvalidOperationException("Base scope should never be popped.");
+			}
+		}
+
+		/// <summary>
+		/// If tracing is enabled, writes an event for the initialization of system-level resources (Generic.xaml etc)
+		/// </summary>
+		internal static IDisposable WriteInitiateGlobalStaticResourcesEventActivity() => _trace.WriteEventActivity(TraceProvider.InitGenericXamlStart, TraceProvider.InitGenericXamlStop);
+
+		/// <summary>
+		/// Register a dictionary for a given source, this is used for retrieval when setting the Source property in code-behind or to an
+		/// external resource.
+		/// </summary>
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public static void RegisterResourceDictionaryBySource(string uri, XamlParseContext context, Func<ResourceDictionary> dictionary)
+		{
+			_registeredDictionariesByUri[uri] = dictionary;
+
+			if (context != null)
+			{
+				// We store the dictionaries inside a ResourceDictionary to utilize the lazy-loading machinery
+				var assemblyDict = _registeredDictionariesByAssembly.FindOrCreate(context.AssemblyName, () => new ResourceDictionary());
+				var initializer = new ResourceDictionary.ResourceInitializer(dictionary);
+				_assemblyRef++; // We don't actually use this key, we just need it to be unique
+				assemblyDict[_assemblyRef] = initializer; 
+			}
+		}
+
+		/// <summary>
+		/// Retrieve the ResourceDictionary mapping to a given source. Throws an exception if none is found.
+		/// </summary>
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public static ResourceDictionary RetrieveDictionaryForSource(Uri source) => RetrieveDictionaryForSource(source.AbsoluteUri, currentAbsolutePath: null);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor this method to add validation of parameter 'source' before using it.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=4826539)